tls-certificates interface (requires side, existing interface)

dshcherb opened the pull request:
This change introduces an implementation of the requires side of
tls-certificates which allows charms to use the CAClient object to
expose inputs for certificate generation by a certificate authority.

Currently easyrsa and Vault charms provide this interface while a client
charm can be any charm that needs to set up TLS termination for its
services.

Certificates and keys in ca_client_test_data.yaml were generated once
for the purposes of adding unit tests and are not used in any of the
running systems.

jameinel commented on test/lib/tls_certificates/test_ca_client.py:
```diff
@@ -0,0 +1,131 @@
+        self.harness.begin()
+        self.ca_client = ca_client.CAClient(self.harness.charm, 'ca-client')
+
+    def test_is_join(self):
```
is_joined

though I think this is as much a test of ca_available as it is of is_joined.

dshcherb replied:
`test_ca_available` might be better then

dshcherb replied:
https://github.com/canonical/operator/pull/210/commits/7407ab5ebd35c1cf52547f399c447e0a6f758bac

jameinel commented on test/lib/tls_certificates/test_ca_client.py:
```diff
@@ -0,0 +1,131 @@
+        # were generated once for the purposes of creating an example.
+        # They are not used anywhere in a production or test system.
+        ca_client_data = yaml.safe_load(
+            (Path(__file__).parent / 'ca_client_test_data.yaml').read_text())
```
Is it cleaner to read a .yaml file or just have it already as a python string? and just refer to it as `test.test_ca_client_data`.
If it is straightforward to generate it as YAML then it makes sense to leave it as such, but if it is just a way to hold test content, I think it is easier as a python script.

dshcherb replied:
OK, it looks like it fits into a plain python file as a string so I will get rid of YAML.

dshcherb replied:
https://github.com/canonical/operator/pull/210/commits/7407ab5ebd35c1cf52547f399c447e0a6f758bac

jameinel commented on test/lib/tls_certificates/test_ca_client.py:
```diff
@@ -0,0 +1,131 @@
+        self.assertTrue(len(receiver.observed_events) == 1)
+        self.assertIsInstance(receiver.observed_events[0], ca_client.TlsConfigReady)
+
+        self.assertIsInstance(self.ca_client.ca_certificate,
```
It seems odd to depend on an instance of a private type here.

dshcherb replied:
An alternative I suggest then is checking some properties of certs and keys based on the public properties that those instances expose.

dshcherb replied:
https://github.com/canonical/operator/pull/210/commits/7407ab5ebd35c1cf52547f399c447e0a6f758bac

dshcherb commented on ops/lib/tls_certificates/ca_client.py:
```diff
@@ -0,0 +1,161 @@
+
+
+from ops.framework import Object, EventBase, EventSetBase, EventSource, StoredState
+from cryptography.hazmat.backends import default_backend
```
True, at the same time only `hazmat` exposes load_pem_private_key AFAIKS:

https://cryptography.io/en/latest/hazmat/primitives/asymmetric/serialization/#cryptography.hazmat.primitives.serialization.load_pem_private_key

My goal with loading and parsing certs/keys is mainly input validation.

dshcherb commented on .travis.yml:
```diff
@@ -18,7 +18,7 @@ matrix:
 
 install:
-    - pip3 install pyyaml autopep8 flake8
+    - pip3 install pyyaml autopep8 flake8 cryptography
```
In a sense, an interface is a python library which may have its own dependencies (including binary).

I think at least a requirements file is needed at the interface level to specify Python dependencies to be downloaded.

As for binary dependencies, we could at least identify a list of things we want available from the target OS distribution. OpenSSL and libyaml are the first two suspects - maybe this is something that could be provided by Juju as a part of preparing a node for running charms in which case we'd have to have a "core" set of binary libs maintained for each target/series (much like with snaps but also for, say, Windows & k8s). Either way, I think we need to move this discussion closer to the packaging spec.

dshcherb replied:
https://github.com/canonical/operator/pull/210/commits/c5389493114ac29127c820cd9bc675665a2e2cd0 (to address the immediate issue of the failing test case).

If this gets merged as is from the dependency handling standpoint, we will have an issue if somebody tries to run tests on their own machine without python3-cryptography available (which will not be the case for contemporary versions of Ubuntu).

dshcherb commented on ops/lib/tls_certificates/ca_client.py:
```diff
@@ -0,0 +1,161 @@
+from cryptography.x509 import load_pem_x509_certificate
+
+import json
+import logging
```
Agreed, thanks for noticing.

dshcherb commented on test/lib/tls_certificates/ca_client_test_data.yaml:
```diff
@@ -0,0 +1,82 @@
+  9bGeQZ4iBP8SjXCxksVQmrGNTE+/ooP/L/e9Jn0f/jOOkJfMKmgAO3PJW03sSwJK
+  vLkurCsHpRm9JzImcopKb+dRBQem9ap74Ab9QX7rUzkH5IKJwjk+wp9fBs9beQBA
+  /topCU469e0vh/yNjo1/h+bdXMWebv4HNAiXPdp6Yi7VohTtyHz2VA==
+  -----END CERTIFICATE-----
```
Yes, it was a dump of `relation-get --format=yaml` - I wonder why it didn't go for a simpler form. I will reformat that.

dshcherb replied:
https://github.com/canonical/operator/pull/210/commits/e1a1bdd97a32e0a9db0e02534020c846058056bc

dshcherb commented on ops/lib/tls_certificates/ca_client.py:
```diff
@@ -0,0 +1,161 @@
+        if cert is None or key is None or ca is None:
+            logger.info('A CA has not yet exposed a requested certificate,'
+                        ' key and CA certificate.')
+            return
```
https://github.com/canonical/operator/pull/210/commits/e1a1bdd97a32e0a9db0e02534020c846058056bc

dshcherb commented on ops/lib/tls_certificates/ca_client.py:
```diff
@@ -0,0 +1,161 @@
+class CAClient(Object):
+
+    on = CAClientEvents()
+    stored = StoredState()
```
https://github.com/canonical/operator/pull/210/commits/e1a1bdd97a32e0a9db0e02534020c846058056bc

dshcherb commented on ops/lib/tls_certificates/ca_client.py:
```diff
@@ -0,0 +1,161 @@
+logger = logging.getLogger(__name__)
+
+
+class CAAvailable(EventBase):
```
https://github.com/canonical/operator/pull/210/commits/e1a1bdd97a32e0a9db0e02534020c846058056bc

dshcherb commented:
Addressed some comments in https://github.com/canonical/operator/pull/210/commits/e1a1bdd97a32e0a9db0e02534020c846058056bc. More changes to come to cover the rest.

I included a source of a fix for https://github.com/canonical/operator/issues/211 (the change with tests is in https://github.com/canonical/operator/pull/212).

dshcherb commented on ops/lib/tls_certificates/ca_client.py:
```diff
@@ -0,0 +1,161 @@
+    def request_server_certificate(self, common_name, sans):
+        """Request a new server certificate.
+
+        If arguments do not change from a previous request, then a new certificate will not
```
ack

dshcherb commented on ops/lib/tls_certificates/ca_client.py:
```diff
@@ -0,0 +1,161 @@
+    @property
+    def key(self):
+        return load_pem_private_key(self.stored.key.encode('utf-8'), password=None,
+                                    backend=default_backend())
```
Once we switch to using controller-side storage, the pattern will change but I agree that as of today the key will end up wherever `/var/lib/juju/agents/unit-<app-name>-<unit-num>/charm/` is mounted which may or may not be a secure rootfs.

Based on my previous experience, charms often store secrets in the leader data (~peer app relation data) and rely on Juju to keep them secure in transit. It's not the case for storage "at rest" but it could be which would solve the problem provided controller-side storage is used.

In my view, we could provide something like `SecureStoredState` albeit it is a very general statement to call something "secure" and will likely give a false sense of security.

jameinel replied:
We've had a few rounds of discussion in Juju about secrets. At one point it was just going to be schema on config. but the more I've thought about it, I like having something separately called Secrets. Mostly because you don't then accidentally log them/write them insecurely because they happen to be part of another data bag. Having it expressly named as a separate thing makes it clearer that they should be filtered.

dshcherb commented:
I addressed the inline comments, I will look at the comments in #216 and see what can be applied here.

dshcherb commented:
@chipaca 

Maybe we need a different label for an existing interface (this one is not new).

dshcherb commented:
@niemeyer 

Hopefully I got all of them in the following commits (apologies for not spotting them earlier):
https://github.com/canonical/operator/pull/210/commits/64b7ef6e9256334757fb80baf4ad80b7f9680bb2
https://github.com/canonical/operator/pull/210/commits/a56fbd98f818c66927f780eae67f23b70b38d9fa
https://github.com/canonical/operator/pull/210/commits/c00cfc7e8afac18b1018864fd70ec0712661a3c1

niemeyer commented:
Closing this as discussed in the standup.  Per discussion in the standup, such interfaces need to evolve inside the charm until they fulfill their role, and until people that can authoritatively speak for such interfaces, as they maintain charms that depend on them for a while, can review the implementation and functionality.

Meanwhile, our recommendation is that the interface should live inside a "lib" submodule of the charm itself, so that it can easily be migrated into the standard library of the framework after that period of review and testing.

stub42 commented:
@niemeyer top level lib/foo ?  Or lib/interface/foo or lib/interfaces/foo or other?
